scripts/idl/README.md

```diff
@@ -0,0 +1,190 @@
+}
+
+
+// On every endpoint number (non-dynamic)
+// a series of clusters can be exposed
+endpoint 0 {
+  // A binding cluster is a CLIENT cluster that can be bound to for the
+  // application to make use.
+  //
+  // As an example, a light switch can be bound to a light bulb or
+  // a cluster can be bound to a OTA provider to use for updates.
+  binding cluster OtaSoftwareUpdateProvider;
+
+  // A server cluster is a server that gets exposed to the world.
+  //
+  // As an example, a light bulb may expose a OnOff cluster.
+  server  cluster OtaSoftwareUpdateRequestor;
+}
+
+```
+
+## Parsing of IDLs
+
+IDL parsing is done within the `idl` python package (this is the current
+directory of this README). Most of the heavy lifting is done by the lark using
+[matter_grammar.lark](./matter_grammar.lark), which is then turned into an AST:
+
+-   [matter_grammar.lark](./matter_grammar.lark) parses and validates textual
+    content
+-   [matter_idl_parser.py](./matter_idl_parser.py) has a transformer that
+    converts the text given by lark into a more type-safe (and type-rich) AST as
+    defined ing [matter_idl_types.py](./matter_idl_types.py)
+
+## Code generation
+
+Code generators are defined in `generators` and their purpose is to convert the
+parsed AST into one or several output files. In most cases the output will be
+split per cluster so that large CPP files are not generated (faster and more
+parallel compilation).
+
+### Code generator base functionality
+
+Generators use [Jinja2](https://jinja.palletsprojects.com/en/3.0.x/) as a
+templating language. The general `CodeGenerator` class in
+[generators/**init**.py](./generators/__init__.py) provides the ability to
+output files based on jinja templates.
+
+In order to build working jinja2 templates, some further processing of the AST
+data is required. Some facilities for lookup namespacing (e.g. search for named
+data types within cluster first then globally) as well interpretation of data
+types into more concrete types is provided by `generators/types.py`.
+
+### Implementing generators
+
+Beyond default AST processing, each generator is expected to add
+language-specific filters to create templates that work. This includes:
+
+-   add any additional filters and transforms for data
+-   add any additional type processing that is language specific
+
+See the java code generator in `generators/java` as an example of codegen.
+
+### Testing generators
+
+Tests of generation are based on checking that a given input matches an expected
+output. The [tests/available_tests](./test/available_tests.yaml) describe for
+each input and generator the expected output.
+
+Intent for tests is to be focused and still easy to see deltas. Input IDLs are
+expected to be small and focusing on a specific functionality. Keep in mind that
+the test outputs are expected to be human-reviwed when codegen logic changes.
+
+These generator tests rare run by `test_generators.py`.
```

Comment:
```suggestion
These generator tests are run by `test_generators.py`.
```

Reply:
This is a neat find. Since rare is a word, spellcheck did not even sniff it out :D Thanks!